### ISSUE TITLE
Impossible to properly create General/kWh with "EnergyMeterMode" = 1 with (at least) legacy Python plugin

### ISSUE THREAD

**Issue by FlyingDomotic** [domoticz/domoticz#6194]
'DomoticzVersion':'2024.7'
'DomoticzHash':'b317cfab1'
'DomoticzBuildTime':'2024-07-13 16:59:31'

Issue : Impossible to properly create General/kWh with "EnergyMeterMode" = 1 with (at least) legacy Python plugin, while ok if "EnergyMeterMode" not defined.

```
"""
<plugin key="Test_kWh_creation" name="Plugin to validate improper kWh device creation" author="Flying Domotic" version="1.0.0">
    <description>
        Plugin to validate improper kWh device creation<br/><br/>
    </description>
    <params>
    </params>
</plugin>
"""
import Domoticz

# Find a device by name in devices table
def getDevice(deviceName):
	for device in Devices:
		if (Devices[device].DeviceID == deviceName) :
			# Return device
			return Devices[device]
	# Return None if not found
	return None

def onStart():
	device1 = Domoticz.Device(Name="EnergyMeterMode1", Unit=1, Type=243, Subtype=29, Switchtype=0, DeviceID="EnergyMeterMode1", Options = {"EnergyMeterMode": 1})
	device1.Create()
	device1.Update(nValue = 0, sValue = "500;0.0")
	device2 = Domoticz.Device(Name="EnergyMeterMode0", Unit=2, Type=243, Subtype=29, Switchtype=0, DeviceID="EnergyMeterMode0")
	device2.Create()
	device2.Update(nValue = 0, sValue = "500;0.0")

	device = getDevice("EnergyMeterMode1")
	Domoticz.Log(F"Device {device.DeviceID}: nValue >{device.nValue}<, sValue >{device.sValue}<")
	device = getDevice("EnergyMeterMode0")
	Domoticz.Log(F"Device {device.DeviceID}: nValue >{device.nValue}<, sValue >{device.sValue}<")
```

Log:
```
2024-10-28 22:53:34.937  Status: TestKwhCreation: Initialized version 1.0.0, author 'Flying Domotic'
2024-10-28 22:53:34.937  Status: Python EventSystem: Initializing event module.
2024-10-28 22:53:34.938  Status: EventSystem: Started
2024-10-28 22:53:34.938  Status: EventSystem: Queue thread started...
2024-10-28 22:53:34.945  TestKwhCreation: Device EnergyMeterMode1: nValue >0<, sValue ><
2024-10-28 22:53:34.945  TestKwhCreation: Device EnergyMeterMode0: nValue >0<, sValue >500;0.0<
```

Probable cause; 
uint64_t CSQLHelper::UpdateValueInt check for sValue with 2 items before making any change. As device is created with null sValue, there's no way to set even a null ("0;0.0") value.

Possible solution:
Create kWh device with a "0;0;0" initial sValue to be able to update it later.

**Comment by gizmocuz** [comment]
Could it be the problem is inside PythonObjects.cpp line 418

```
		else if (sTypeName == "kWh")
		{
			sValue = "0; 0.0";
			SubType = sTypeKwh;
		}
```

The space behind the semicolum is also strange, maybe it should be

```
		else if (sTypeName == "kWh")
		{
			sValue = "0;0;0";
			SubType = sTypeKwh;
		}
```

I am having issues using the code above, it does not show up under the hardware page. I do see other python plugins here

When you do not call 'device.Update', what's the sValue?

**Comment by FlyingDomotic** [comment]
This could also be part of the issue, but perhaps when using sTypeName (which is 
not my case, I use Type and SubType with numerical values).

In addition, sValue is left empty after creation (as you can see in log).

You probably mean

```
sValue = "0;0.0";

```
in the proposed fix?

Code shows up (after restarting Domoticz) under "Plugin to validate improper kWh device creation" at home.

Le 29/10/2024 à 07:42, gizmocuz a écrit :
> Could it be the problem is inside PythonObjects.cpp line 418
>
> ```
> 		else if (sTypeName == "kWh")
> 		{
> 			sValue = "0; 0.0";
> 			SubType = sTypeKwh;
> 		}
> ```
>
> The space behind the semicolum is also strange, maybe it should be
>
> ```
> 		else if (sTypeName == "kWh")
> 		{
> 			sValue = "0;0;0";
> 			SubType = sTypeKwh;
> 		}
> ```
>
> I am having issues using the code above, it does not show up under the hardware page. I do see other python plugins here
>
> When you do not call 'device.Update', what's the sValue?
>
>
>

**Comment by FlyingDomotic** [comment]
Another (perhaps more generic) solution could be adding nValue and sValue as optionnal parameters to init device, allowing to bypass checks (as does maptypename) and give only one call when needing to initialize n/sValues

**Comment by gizmocuz** [comment]
when you add a 'dummy' hardware and create a kWh sensor, it calls sql.CreateDevice, and also contains the right sValue.
I am not into the python code, but when you call 'create' with the python api, the same should happen

**Comment by FlyingDomotic** [comment]
C code and Python code don't use the same primitives, thus parameters may be (are) different.
Problem here is that Python .create primitive don't allow nValue/sValue to be set when creating the device. Then, when trying to update sValue, a check is done when "EnergyMeterMode" is set to one, and update is discarded if previous sValue has not 2 values separated by a ";" (which is the case when created with null/empty value). And as soon as sValue is invalid, there's no way to fix it (except updating manually database, not a good idea indeed).
There are (at least) 2 ways to fix that:
1. Allow create to set initlal sValue, or force initial value to "0;0.0" when creating the device, even with numerical values.
2. "Forcing" sValue" to "0;0.0" when reading a device with previous sValue not defined.
Note that with "EnergyMeterMode" different of 1, there's not problem, as given sValue is overwritting previous one.

**Comment by Mike19700** [comment]
I am currently working on a plugin for my PowerWorld heat pump and am running into the same problem.
I am creating a device 'Compressor power:
            Opt={'EnergyMeterMode': '1' }
            Domoticz.Device(Name=“Compressor power”,Unit=21,Type=243,Subtype=29,Options=Opt,Used=1).Create()

Next, I enter data:
Devices[21].Update(nValue=123, sValue=“{661;123.123}”)
but the device shows nothing.
![Schermafbeelding 2024-11-10 134724](https://github.com/user-attachments/assets/af11031b-3a94-404f-9e51-6f2b6dc70365)

If I set the device to 'Energy read' from device and I enter the same data , it does shows the calculated data in the device however not the current powerusage.
![Schermafbeelding 2024-11-10 135203](https://github.com/user-attachments/assets/cfd6dde3-5c36-4694-947e-0f9ce0e27353)

**Comment by gizmocuz** [comment]
Maybe best to go to our wiki and read the documentation about the JSON API. Then you understand what/how to update the devices and what the sValue meaning is.

**Comment by Mike19700** [comment]
> Maybe best to go to our wiki and read the documentation about the JSON API. Then you understand what/how to update the devices and what the sValue meaning is.
> 

Normaly I use the Json API but early this year I had the same problem with a script I wrote and thought it was my own fault. The last 2 days I really got frustrated because it did not work the way it is described at the wiki page how to build a plugin. The I saw that someone else got the same problem as I do. So I thought maybe it is not my fault. Please believe me that I only post a problem here if I already spend hours to solve it. I do not understand why I have to go to the json wiki that I already know very good. I think there could be a program inrerpretation I do not understand. Would anybody be so kind to explain what I am doing wrong in my example?

**Comment by FlyingDomotic** [comment]
As told, and even after further investigation, it's impossible to create "General/Kwh" device with "EnergyMeterMode" = 1 with a Python plugin, while it works perfectly with C code.
Reason is a too strict control on sValue, which can't be changed if illegal, and is created illegal (empty) by code behind Python plugin .create primitive.
Fixing can only be in C code, either by setting initial sValue properly (fpr example as "0;0.0", or making controls on change more permissive (at least accepting the default value created by C code behind .create from Python), or simply returning "0;0.0" when an illegal value is found.

**Comment by gizmocuz** [comment]
I already mentioned above that you should create a dummy kWh sensor and look at the sValue.
I also pointed out where (probably) in the c++ code the problem lies in relationship to the created kWh sensor in Python.
Maybe it's time to make a patch and build Domoticz yourself? It's explained on our Wiki how to do this and not complicated at all.
If you can program in Python, you should be able to change a few lines and follow compile instructions

https://www.domoticz.com/wiki/Build_Domoticz_from_source

**Comment by FlyingDomotic** [comment]
Thanks for your advice!

I'll try to compile a version without the space in a first time, and propose the patch as PR (as I mostly don't need this change for my own, but user's of my plugins do ;-).

I would when propose adding initial nValue/sValue to python device.create function as optional parameters. As described in "contrib.md" at https://github.com/domoticz/domoticz/blob/development/.github/CONTRIBUTING.md, I need first to ask for your validation about this new feature/change. Would this addition to the plugin framework be ok for you?

**Comment by gizmocuz** [comment]
That sound great! But when these parameters are not supplied or are invalid, please make sure the device is created properly!
For instance if a kWh sensor needs 3 sValue field x;y;z, and the user supplies one,two or sixty, create a standard kWh device with 3 with the default values, the same as sql.CreateDevice does
Maybe make use of that function?

**Comment by Mike19700** [comment]
> Thanks for your advice!
> 
> I'll try to compile a version without the space in a first time, and propose the patch as PR (as I mostly don't need this change for my own, but user's of my plugins do ;-).
> 
> I would when propose adding initial nValue/sValue to python device.create function as optional parameters. As described in "contrib.md" at https://github.com/domoticz/domoticz/blob/development/.github/CONTRIBUTING.md, I need first to ask for your validation about this new feature/change. Would this addition to the plugin framework be ok for you?

Tried to compile a version to but have a problem to install make. Otherwise I would be glad to help.

**Comment by FlyingDomotic** [comment]
Some news:

I made suggested change into PythonObjects.cpp, and compiled it.

Test plugin still fails.

Reason is probably that the change was into maptypename function, which seems to be called only when a typeName is provided (which is not the case when using only numerical value for type and subType).

Anyway, I'll do a PR to fix the issue when creation is done through typeName.

**Comment by pipiche38** [comment]
I'm wondering if the error is not in the code

`device1 = Domoticz.Device(Name="EnergyMeterMode1", Unit=1, Type=243, Subtype=29, Switchtype=0, DeviceID="EnergyMeterMode1", Options = {"EnergyMeterMode": 1})`

My understanding is that 'EnergyMeterMode' should be a string and not an int, as for the zigbee4domoticz plugin we are doing 

`Options = {'EnergyMeterMode': '0'} ` for (from Device) 
or 
`Options = {'EnergyMeterMode': '1'}`  for from Compute

and we do not have any issue to report when creating the wdiget

**Comment by WillemD61** [comment]
The bug still exists and it definitely is a bug. (I am using the extended plugin framework, not the legacy, but the problem is the same)

I can created the device using the python plugin and set it to "computed", no problem. It looks the same as other devices of the same type.

But I cannot update the device with the plugin if there is no initial svalue present. Below is the code and the output of the code showing the problem.

Now if I put an initial svalue on the device (with sqlite 3 directly on the devicestatus table), then after that the update with the plugin works without problems. So either the python plugin update needs to be fixed or the intial creation of the device needs to put an svalue on it.

I have posted my findings in the forum
[(https://forum.domoticz.com/viewtopic.php?t=43758)]

More info: also updating the device with a JSON call does not work if there is no initial sValue present. So the problem really is in the update processing. The only way to work around this is to manual add a svalue directly in the dbase first.

**Comment by WillemD61** [comment]
I think I found where the problem is. It is in line 5371-5397 in the SQLhelper.cpp file.

If the initial svalue does not consists of two elements then it just restores the initial value. So if the initial value is empty (which it always is after the device is created) then the svalue remains empty. Here is the offending code.

```
            std::vector<std::string> powerAndEnergyBeforeUpdate;
			StringSplit(sValueBeforeUpdate, ";", powerAndEnergyBeforeUpdate);
			if (powerAndEnergyBeforeUpdate.size() == 2)
			{
				//we need to use atof here because some users seem to have a illegal sValue in the database that causes std::stof to crash
				double powerDuringInterval = atof(powerAndEnergyBeforeUpdate[0].c_str());
				double energyUpToInterval = atof(powerAndEnergyBeforeUpdate[1].c_str());
				double energyDuringInterval = powerDuringInterval * intervalSeconds / 3600;
				double energyAfterInterval = energyUpToInterval + energyDuringInterval;
				std::vector<std::string> powerAndEnergyUpdate;
				StringSplit(sValue, ";", powerAndEnergyUpdate);
				if (!powerAndEnergyUpdate.empty())
				{
					const char* powerUpdate = powerAndEnergyUpdate[0].c_str();
                    char sValueUpdate[100];
                    sprintf(sValueUpdate, "%s;%.4f", powerUpdate, energyAfterInterval);
					sValue = sValueUpdate;
				}
				else
				{
                    sValue = sValueBeforeUpdate.c_str();
				}
			}
			else
            {
                sValue = sValueBeforeUpdate.c_str();
            }
```

**Comment by WillemD61** [comment]
I found a workaround

1) First I create the kWh device using the plugin with the default "from device" setting (leave the options field empty when creating the device).
2) Immediately after creation, I update the device svalue to "0;0" using the plugin.
3) and then immediately I update the options field to 'EnergyMeterMode': '1' which changes the device to "computed" instead of "from device".

Note in step 3 the update statement needs to be run with the UpdateOptions=True flag like this
Devices[DeviceID].Units[Unit].Update(UpdateOptions=True)

Now it has an initial value and future updates will work.

**Comment by gizmocuz** [comment]
That code above assumes there are two sValues, which there should be.
This is not the problem code.

The problem code is the code that the Python plugin system uses to create the device.
Some devices need to have even more separated values in the sValue

If you want to know how to create a correct device, it is best to see the code that creates a new dummy sensor (from the web gui)

**Comment by WillemD61** [comment]
Why does the update of an initial sValue work correctly on a "from device" device but not on a "computed" device? The code referenced explains why the update of the computed kWh device does not work. 

As you say, in order for the code to work, sValue should be of structure "0;0", but it is not since the device is created with empty sValue.

It could indeed be solved if the code that creates the device puts an initial sValue on the device in the format "0;0". That would be addressing the root cause. I don't know where that code is. The plugin framework itself does not allow specification of an sValue during creation.

The problem can be solved in 3 ways now:
1) create the device with an sValue of the correct structure (when the create is called from the plugin)
2) adapt the referenced code that updates the device so that when the sValue on the device is empty it will update the device with the sValue supplied during the update.
3) with the workaround : first create a "from device" device, update the sValue, then change the device to a "computed" device

**Comment by gizmocuz** [comment]
I think you have to ask yourself also the question, why, when you create a new sensor for a dummy hardware (from the hardware page), it is correct...

So the Python plugin code should detect this.

I am also not a fan of the way the Python code creates devices. As you see when you create a new sensor for a dummy device, you can not select any of the 10 types of temp+hum sensors, there is just one.

But a simple piece of code that checks what type of device the plugin wants to create, and creates default values, and when not specified uses these (sValue), that would solve the issue.
I don't think plugin developers want a mandatory sValue in the create call

Next, you could also make a check function (if the above function is created) that checks if the correct amount of parameters are filled in when updating a sensor via Python

You can see the amount of parameters in BasePush.cpp, get the string from the table, split it by a comma , and you have the count

**Comment by WillemD61** [comment]
> I think you have to ask yourself also the question, why, when you create a new sensor for a dummy hardware (from the hardware page), it is correct...

Exactly, and when you test this, you will see that when you created a kWh device from the hardware page, it is created as "from device" and it is populated with an initial sValue "0;0". So when you then change it to "computed" it already has an sValue and therefore any further updates are done without problems. (like the workaround does)

I don't know who maintains the plugin create() function code but if that can be adapted to function in the same way, so add an initial sValue, then the problem is solved.

**Comment by waltervl** [comment]
@WillemD61 @FlyingDomotic 
I did not see you tried the below mentioned solution @pipiche38 is using in its Zigbee4Domoticz Python plugin where the code runs without issues (according him ;-) )

> I'm wondering if the error is not in the code
> 
> `device1 = Domoticz.Device(Name="EnergyMeterMode1", Unit=1, Type=243, Subtype=29, Switchtype=0, DeviceID="EnergyMeterMode1", Options = {"EnergyMeterMode": 1})`
> 
> My understanding is that 'EnergyMeterMode' should be a string and not an int, as for the zigbee4domoticz plugin we are doing
> 
> `Options = {'EnergyMeterMode': '0'} ` for (from Device) or `Options = {'EnergyMeterMode': '1'}` for from Compute
> 
> and we do not have any issue to report when creating the wdiget

**Comment by FlyingDomotic** [comment]
I agree, putting 'EnergyMeterMode' value as string will probably solve the problem

**Comment by WillemD61** [comment]
No, it does not, I already had it as a string.

**Comment by FlyingDomotic** [comment]
For this kind of change, that have an impact at device creation, you must first delete device and then restart MqttMapper (od Domoticz) to let it created from scratch.

Did you proceed this way?

**Comment by WillemD61** [comment]
yes, for each test I removed all devices and started all over again.

**Comment by FlyingDomotic** [comment]
I just check the code, "options" value is send as is to Domoticz create device function.

Would it be possible to get a copy and paste of this part from config file?

**Comment by waltervl** [comment]
> I'm wondering if the error is not in the code
> 
> `device1 = Domoticz.Device(Name="EnergyMeterMode1", Unit=1, Type=243, Subtype=29, Switchtype=0, DeviceID="EnergyMeterMode1", Options = {"EnergyMeterMode": 1})`
> 
> My understanding is that 'EnergyMeterMode' should be a string and not an int, as for the zigbee4domoticz plugin we are doing
> 
> `Options = {'EnergyMeterMode': '0'} ` for (from Device) or `Options = {'EnergyMeterMode': '1'}` for from Compute
> 
> and we do not have any issue to report when creating the wdiget

@pipiche38 can you explain what method you use in the zigbee4domoticz plugin to create the device and send sValue data to an energy device with 'EnergyMeterMode': '1'}?

**Comment by WillemD61** [comment]
If you want to see my code as well, here is the relevant extract from my plugin code, including the workaround.

I have a dictionary of devices to be created (showing only 2 now).
onStart creates the devices if they don't exists already.
Any device is created as defined, only device of type 243, subtype 29 is created without options field first (i.e. "from device" kWh device) and then updated to device including the options field (i.e. "computed" kWh device)

```
DEVSLIST={
# 243,29,0 or 4: kwh device
    "50f2":  [ 43, 243, 29, 0, {'EnergyMeterMode': '1'}, 1, "HP energy usage"],
    "5077":  [ 44, 243, 29, 4, {'EnergyMeterMode': '1'}, 1, "HP energy generated"],
}


    def onStart(self):
        Domoticz.Log("onStart called")
        Domoticz.Heartbeat(int(Parameters["Mode1"]))
        self.Hwid=Parameters['HardwareID']
        # cycle through device list and create any non-existing devices when the plugin/domoticz is started
        for Dev in DEVSLIST:
            Unit=DEVSLIST[Dev][0]
            DeviceID="{:04x}{:04x}".format(self.Hwid,Unit)
            Type=DEVSLIST[Dev][1]
            Subtype=DEVSLIST[Dev][2]
            Switchtype=DEVSLIST[Dev][3]
            Options=DEVSLIST[Dev][4]
            Name="XTEND: "+DEVSLIST[Dev][6]
            Description="Xtend field code :"+Dev
            if DeviceID not in Devices:
                Domoticz.Status(f"Creating device for Field {Dev} ...")
                if ((Type==243) and (Subtype==29)):
                    # below code puts an initial svalue on the kwh device and then changes the type to "computed". This is to work around a BUG in Domoticz for computed kwh devices. See issue 6194 on Github.
                    Domoticz.Unit(DeviceID=DeviceID,Unit=Unit, Name=Name, Type=Type, Subtype=Subtype, Switchtype=Switchtype, Options={}, Used=1, Description=Description).Create()
                    Devices[DeviceID].Units[Unit].sValue="0;0"
                    Devices[DeviceID].Units[Unit].Update()
                    Devices[DeviceID].Units[Unit].Options=Options
                    Devices[DeviceID].Units[Unit].Update(UpdateOptions=True)
                else:
                    Domoticz.Unit(DeviceID=DeviceID,Unit=Unit, Name=Name, Type=Type, Subtype=Subtype, Switchtype=Switchtype, Options=Options, Used=1, Description=Description).Create()
```

**Comment by gizmocuz** [comment]
I know this is quite an old thread, but is this still an issue or is this already solved in the latest beta version?

**Comment by WillemD61** [comment]
I use latest stable, not beta, and I use the workaround as indicated above in my plugin, so I am not aware of any solution. Sorry.

**Comment by gizmocuz** [comment]
OK, so it is probably solved? Maybe try the beta? (a backup is made when you issue ./updatebeta, but you can always stop domoticz, and make a backup of the complete folder). Or if you use docker, stop it, make a backup of the volume, update to 'latest'

**Comment by WillemD61** [comment]
I will test within a few days.

**Comment by WillemD61** [comment]
I did the test today on latest beta (Domoticz V2025.2 (build 17099)) and I conclude that the problem still exists.

I created a minimal testplugin for the test which does the following:
1) onStart it creates two kWh devices, one using the workaround and one without the workaround.
2) onHeartbeat it tries to update those two devices with a value that is incremented on each heartbeat.

The device that was created with the workaround is update without problems. The other one remains empty.

The minimal code for the testplugin is here, so anyone can test this in an easy way by just installing and activating this testplugin.

```
"""
<plugin key="testPlugin" name="testplugin" author="WillemD61" version="1.0.7" >
    <description>
        <h2>testplugin for electricity devices</h2><br/>
    </description>
    <params>
        <param field="Mode1" label="Polling Interval" width="150px">
            <options>
                <option label="10 seconds" value="10" />
                <option label="20 seconds" value="20" />
                <option label="30 seconds" value="30" default="true" /> # maximum domoticz heartbeat time is 30 seconds
                <option label="1 minute" value="60" />
                <option label="2 minutes" value="120" />
                <option label="3 minutes" value="180" />
                <option label="4 minutes" value="240" />
                <option label="5 minutes" value="300" />
            </options>
        </param>
    </params>
</plugin>
"""
import DomoticzEx as Domoticz
import json,requests   # make sure these are available in your system environment
from requests.exceptions import Timeout

class testPlugin:
    enabled = False
    def __init__(self):
        return

    def onStart(self):
        Domoticz.Log("onStart called with parameters")
        for elem in Parameters:
            Domoticz.Log(str(elem)+" "+str(Parameters[elem]))
        if int(Parameters["Mode1"])<=30:
            Domoticz.Heartbeat(int(Parameters["Mode1"]))
            self.heartbeatWaits=0
        else:
            Domoticz.Heartbeat(30)
            self.heartbeatWaits=int(int(Parameters["Mode1"])/30 - 1)
        self.heartbeatCounter=0
        self.Hwid=Parameters['HardwareID']
        self.testCounter=0

        # create two devices , one with workaround and one without
        Unit=1
        DeviceID="{:04x}{:04x}".format(self.Hwid,Unit)
        Type=243
        Subtype=29
        Switchtype=0
        Options={'EnergyMeterMode': '1'}
        Name="with workaround"
        Description="test device with workaround"
        if DeviceID not in Devices:
            Domoticz.Status(f"Creating device with workaround ...")
            Domoticz.Unit(DeviceID=DeviceID,Unit=Unit, Name=Name, Type=Type, Subtype=Subtype, Switchtype=Switchtype, Options={}, Used=1, Description=Description).Create()
            Devices[DeviceID].Units[Unit].sValue="0;0"
            Devices[DeviceID].Units[Unit].Update()
            Devices[DeviceID].Units[Unit].Options=Options
            Devices[DeviceID].Units[Unit].Update(UpdateOptions=True)

        Unit=2
        DeviceID="{:04x}{:04x}".format(self.Hwid,Unit)
        Type=243
        Subtype=29
        Switchtype=0
        Options={'EnergyMeterMode': '1'}
        Name="without workaround"
        Description="test device without workaround"
        if DeviceID not in Devices:
            Domoticz.Status(f"Creating device without workaround ...")
            Domoticz.Unit(DeviceID=DeviceID,Unit=Unit, Name=Name, Type=Type, Subtype=Subtype, Switchtype=Switchtype, Options=Options, Used=1, Description=Description).Create()

    def onStop(self):
        Domoticz.Log("onStop called")

    def onConnect(self, Connection, Status, Description):
        Domoticz.Log("onConnect called")

    def onMessage(self, Connection, Data):
        Domoticz.Log("onMessage called")

    def onCommand(self, DeviceID, Unit, Command, Level, Color):
        Domoticz.Log("onCommand called for Device " + str(DeviceID) + " Unit " + str(Unit) + ": Parameter '" + str(Command) + "', Level: " + str(Level))

    def onNotification(self, Name, Subject, Text, Status, Priority, Sound, ImageFile):
        Domoticz.Log("Notification: " + Name + "," + Subject + "," + Text + "," + Status + "," + str(Priority) + "," + Sound + "," + ImageFile)

    def onDisconnect(self, Connection):
        Domoticz.Log("onDisconnect called")

    def onHeartbeat(self):
        Domoticz.Log("onHeartbeat called")
        if self.heartbeatWaits==self.heartbeatCounter:
            self.testCounter+=1
            Unit=1
            DeviceID="{:04x}{:04x}".format(self.Hwid,Unit)
            Devices[DeviceID].Units[Unit].nValue=0
            Devices[DeviceID].Units[Unit].sValue=str(self.testCounter)+";1" # watts are supplied, kwh are calculated by Domoticz.
            Devices[DeviceID].Units[Unit].Update()

            Unit=2
            DeviceID="{:04x}{:04x}".format(self.Hwid,Unit)
            Devices[DeviceID].Units[Unit].nValue=0
            Devices[DeviceID].Units[Unit].sValue=str(self.testCounter)+";1" # watts are supplied, kwh are calculated by Domoticz.
            Devices[DeviceID].Units[Unit].Update()

            self.heartbeatCounter=0
        else:
            self.heartbeatCounter+=1

global _plugin
_plugin = testPlugin()

def onStart():
    global _plugin
    _plugin.onStart()

def onStop():
    global _plugin
    _plugin.onStop()

def onConnect(Connection, Status, Description):
    global _plugin
    _plugin.onConnect(Connection, Status, Description)

def onMessage(Connection, Data):
    global _plugin
    _plugin.onMessage(Connection, Data)

def onCommand(DeviceID, Unit, Command, Level, Color):
    global _plugin
    _plugin.onCommand(DeviceID, Unit, Command, Level, Color)

def onNotification(Name, Subject, Text, Status, Priority, Sound, ImageFile):
    global _plugin
    _plugin.onNotification(Name, Subject, Text, Status, Priority, Sound, ImageFile)

def onDisconnect(Connection):
    global _plugin
    _plugin.onDisconnect(Connection)

def onHeartbeat():
    global _plugin
    _plugin.onHeartbeat()

# Generic helper functions
def DumpConfigToLog():
    for x in Parameters:
        if Parameters[x] != "":
            Domoticz.Debug( "'" + x + "':'" + str(Parameters[x]) + "'")
    Domoticz.Debug("Device count: " + str(len(Devices)))
    for DeviceName in Devices:
        Device = Devices[DeviceName]
        Domoticz.Debug("Device ID:       '" + str(Device.DeviceID) + "'")
        Domoticz.Debug("--->Unit Count:      '" + str(len(Device.Units)) + "'")
        for UnitNo in Device.Units:
            Unit = Device.Units[UnitNo]
            Domoticz.Debug("--->Unit:           " + str(UnitNo))
            Domoticz.Debug("--->Unit Name:     '" + Unit.Name + "'")
            Domoticz.Debug("--->Unit nValue:    " + str(Unit.nValue))
            Domoticz.Debug("--->Unit sValue:   '" + Unit.sValue + "'")
            Domoticz.Debug("--->Unit LastLevel: " + str(Unit.LastLevel))
    return


```

**Comment by gizmocuz** [comment]
Should work now in beta 17106 which should be available in an hour

<img width="991" height="168" alt="Image" src="https://github.com/user-attachments/assets/eddff83d-30ef-4a05-8630-5538e6404a31" />

**Comment by WillemD61** [comment]
I confirm, it works

**Comment by gizmocuz** [comment]
Thanks!
